src/passes/JSPI.cpp

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "asmjs/shared-constants.h"
+#include "ir/element-utils.h"
+#include "ir/import-utils.h"
+#include "ir/literal-utils.h"
+#include "ir/utils.h"
+#include "pass.h"
+#include "shared-constants.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+#include <utility>
+
+//
+// Convert a module to be compatible with JavaScript promise integration (JSPI).
+// All exports will be wrapped with a function that will handle storing
+// the suspsender that is passed in as the first param from a "promising"
+// `WebAssembly.Function`. All imports will also be wrapped, but they will take
+// the stored suspender and pass it as the first param to the imported function
+// that should be created from a "suspending" `WebAssembly.Function`.
+//
+namespace wasm {
+
+struct JSPI : public Pass {
+
+  JSPI() {}
+
+  Type externref = Type(HeapType::ext, Nullable);
+
+  void run(PassRunner* runner, Module* module) override {
+    Builder builder(*module);
+    // Create a global to store the suspender that is passed into exported
+    // functions and will then need to be passed out to the imported functions.
+    module->addGlobal(builder.makeGlobal(Name("suspender"),
```

Comment:
Is there common way of naming globals so it doesn't conflict?

Reply:
Yes, see `getValid*Name()` in `src/ir/names.h`.